pkg/controller/kubedirectorconfig/kubedirectorconfig_controller.go

```diff
@@ -0,0 +1,106 @@
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+var log = logf.Log.WithName("controller_kubedirectorconfig")
+
+// Add creates a new KubeDirectorConfig Controller and adds it to the Manager. The Manager will set fields on the Controller
+// and Start it when the Manager is Started.
+func Add(mgr manager.Manager) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	return &ReconcileKubeDirectorConfig{Client: mgr.GetClient(), scheme: mgr.GetScheme()}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("kubedirectorconfig-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	// Watch for changes to primary resource KubeDirectorConfig
+	err = c.Watch(&source.Kind{Type: &v1alpha1.KubeDirectorConfig{}}, &handler.EnqueueRequestForObject{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileKubeDirectorConfig implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileKubeDirectorConfig{}
+
+// ReconcileKubeDirectorConfig reconciles a KubeDirectorConfig object
+type ReconcileKubeDirectorConfig struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	Client client.Client
+	scheme *runtime.Scheme
+	lock         sync.RWMutex
+	globalConfig *v1alpha1.KubeDirectorConfig
+}
+
+// Reconcile reads that state of the cluster for a KubeDirectorConfig object and makes changes based on the state read
+// and what is in the KubeDirectorConfig.Spec
+// Note:
+// The Controller will requeue the Request to be processed again if the returned error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (r *ReconcileKubeDirectorConfig) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling KubeDirectorConfig")
+
+	// Fetch the KubeDirectorConfig instance
+	kdConfig := &v1alpha1.KubeDirectorConfig{}
+	err := r.Client.Get(context.TODO(), request.NamespacedName, kdConfig)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
```

Comment:
So just to confirm... are we always going to get a reconcile request when deletion happens? I.e. this case is expected when deletion happens (and is not just from a modify/deletion race).

Reply:
From the delete testing I've done I believe that's the case. Reconcile got called twice on every delete, once where the Get succeeded and then when the Get returned IsNotFound.